### PR TITLE
Fix interop.reactivestreams.StreamUnicastPublisher

### PR DESCRIPTION
**Pull request by BalmungSan**
Improves the current implementation of `interop.reactivestreams.StreamUnicastPublisher` to allow it to be subscribed by multiple `Publishers` and gracefully cancel active `Subscriptions` when the `Resource` is closed.

-----

**Roadmap**.

- [x] Base implementation.
- [x] Fix tests.
- [x] Backport to the **Java** `Flow` PR.
- [x] Fix MiMa.

-----

Related to #3102

### CODE REVIEW

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 1]
```diff
@@ -67,57 +71,75 @@ private[reactivestreams] final class StreamSubscription[F[_], A](
...
-      else if (n > 0) (Finite(n): Request).pure[F]
-      else F.raiseError(new IllegalArgumentException(s"3.9 - invalid number of elements [$n]"))
-
-    val prog = cancelled.get
-      .ifM(ifTrue = F.unit, ifFalse = request.flatMap(requests.offer).handleErrorWith(onError))
+  private def cancelMe: F[Unit] =
+    F.delay(completed.set(true)) >> cancelled.set(true)
+  override def cancel(): Unit =
+    if (!completed.get()) {
+      requestDispatcher.unsafeRunSync(cancelMe)
```
[the remark this answers is not in the document]
Comment (BalmungSan) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:94]: @armanbilge Mostly to ensure future `cancel` and `request` calls to be NOOPs. Thus I ensure that `subscribe` waits for the `completed` and `cancelled` flags to be set.
And since those two operations are pretty fast, we are not imposing a hard block on our callers.

[the remark this answers is not in the document]
Comment (armanbilge) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:94]: > Mostly to ensure future `cancel` and `request` calls to be NOOPs. Thus I ensure that `subscribe` waits for the `completed` and `cancelled` flags to be set.

Thanks, but it's not obvious to me why guaranteeing this behavior this requires running this action synchronously. According to the comment the spec says it may run asynchronously.

---

> And since those two operations are pretty fast, we are not imposing a hard block on our callers.

`unsafeRunSync()` always forces a shift across threadpools and possibly creating a new thread for blocking. It's hardly a trivial operation.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 2]
```diff
@@ -67,57 +71,75 @@ private[reactivestreams] final class StreamSubscription[F[_], A](
...
+  override def request(n: Long): Unit =
+    if (!completed.get()) {
+      val prog =
+        if (n == java.lang.Long.MAX_VALUE)
+          requests.offer(Infinite)
+        else if (n > 0)
+          requests.offer(Finite(n))
+        else
+          onError(new IllegalArgumentException(s"Invalid number of elements [${n}]"))
+      requestDispatcher.unsafeRunSync(prog)
```
Comment (armanbilge) [inline reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:106]: This one also changed.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala [hunk 3]
```diff
@@ -25,35 +25,43 @@ package reactivestreams
...
+      startDispatcher.unsafeRunAndForget(
+        StreamSubscription.subscribe(stream, subscriber)
+      )
+    catch {
+      case _: IllegalStateException =>
+        subscriber.onSubscribe(new Subscription {
+          override def cancel(): Unit = ()
+          override def request(x$1: Long): Unit = ()
+        })
+        subscriber.onError(StreamUnicastPublisher.CanceledStreamPublisherException)
```
Comment (armanbilge) [inline reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala:64]: Why is this calling `.onSubscribe` first? The Javadoc for `Publisher` says this:

> If already subscribed, or the attempt to subscribe fails due to policy violations or errors, the `Subscriber`'s `onError` method is invoked with an `IllegalStateException`. Otherwise, the `Subscriber`'s `onSubscribe` method is invoked

https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.Publisher.html#subscribe-java.util.concurrent.Flow.Subscriber-

So it seems unnecessary.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (armanbilge) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:94]: Is the `completed.get()` intended to guard against invalid invocations of the `Dispatcher`? If so, I think there is a race condition here with concurrent invocations of `cancel()` and `request()`.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala (hunk 3 above)
Comment (BalmungSan) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala:64]: @armanbilge specification `9` for `Publisher`: https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md

> Publisher.subscribe MUST call onSubscribe on the provided Subscriber prior to any other signals to that Subscriber and MUST [return normally](https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md#term_return_normally), except when the provided Subscriber is null in which case it MUST throw a java.lang.NullPointerException to the caller, for all other situations the only legal way to signal failure (or reject the Subscriber) is by calling onError (after calling onSubscribe).

Actually, I added this because a test failed because of that.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (BalmungSan) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:94]: > Thanks, but it's not obvious to me why guaranteeing this behavior this requires running this action synchronously. According to the comment the spec says it may run asynchronously.

Yes, it may run asynchronously.
However, if a `Subscriber` does something like:

```scala
subscription.cancel()
subscription.request()
```

That `request` HAS to be a NOOP, and if we don't block the `cancel` to set `completed` before returning, then it may not be a NOOP.
I don't remember, but I think I had a failing test because of that.

-----

> unsafeRunSync() always forces a shift across threadpools and possibly creating a new thread for blocking.

I was under the impression that shouldn't be the case with a `sequential` `Dispatcher`, since AFAIK, those were optimized for this very use case, running something simple like `Queue.offer` synchronously.

-----

> Is the completed.get() intended to guard against invalid invocations of the Dispatcher?

Yes, although is a bit more general.
The idea is that `completed` is set when we reach what the spec calls a _"terminal state"_, after which all operations are NOOPs.

> If so, I think there is a race condition here with concurrent invocations of cancel() and request().

😿
That is probably the bug we are seeing, right?

[the remark this answers is not in the document]
Comment (armanbilge) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:94]: > However, if a `Subscriber` does something like:
> 
> ```scala
> subscription.cancel()
> subscription.request()
> ```
> 
> That `request` HAS to be a NOOP, and if we don't block the `cancel` to set `completed` before returning, then it may not be a NOOP.

This ordering will be preserved by the sequential dispatcher :) so we just need to check the condition in the submitted effect, and no-op it if appropriate.

---

> I was under the impression that shouldn't be the case with a `sequential` `Dispatcher`, since AFAIK, those were optimized for this very use case, running something simple like `Queue.offer` synchronously.

No, I'm afraid that's fundamental: you are asynchronously submitting an effect to the `IORuntime` from possibly outside the runtime, and blocking while you await its completion.

The optimization of a sequential dispatcher is that it runs all submitted effects on the same fiber, which is why it can only run one effect at a time and preserves order of (non-concurrent) submission.

The parallel dispatcher creates a new fiber for each submitted effect, which is overhead for simple operations as you point out.

---

> That is probably the bug we are seeing, right?

Possibly 😅 is there a situation when the `Dispatcher` is shutdown that we _shouldn't_ no-op? Seems like we could just ignore this exception.

File: build.sbt [hunk 4]
```diff
@@ -174,6 +174,11 @@ ThisBuild / mimaBinaryIssueFilters ++= Seq(
   ),
   ProblemFilters.exclude[DirectMissingMethodProblem](
     "fs2.io.net.tls.S2nConnection#RecvCallbackContext.readBuffer"
+  ),
+  // Private cosntructor.
+  // Added in #3107
+  ProblemFilters.exclude[DirectMissingMethodProblem](
+    "fs2.interop.reactivestreams.StreamUnicastPublisher.this"
```
Comment (BalmungSan) [inline build.sbt:181]: This constructor was `private` so I think the exclusion is right.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala [hunk 5]
```diff
@@ -25,35 +25,36 @@ package reactivestreams
...
-final class StreamUnicastPublisher[F[_]: Async, A] private (
+final class StreamUnicastPublisher[F[_]: Async, A](
     val stream: Stream[F, A],
-    startDispatcher: Dispatcher[F],
-    requestDispatcher: Dispatcher[F]
+    startDispatcher: Dispatcher[F]
 ) extends Publisher[A] {
-
-  @deprecated("Use StreamUnicastPublisher.apply", "3.4.0")
-  def this(stream: Stream[F, A], dispatcher: Dispatcher[F]) = this(stream, dispatcher, dispatcher)
```
Comment (BalmungSan) [inline reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala:45]: So, the new constructor is the same as the previously deprecated overload, so I just make it public to satisfy **MIMA**.
Is this correct?

File: reactive-streams/src/test/scala/fs2/interop/reactivestreams/CancellationSpec.scala [hunk 6]
```diff
@@ -36,54 +35,44 @@ import java.util.concurrent.atomic.AtomicBoolean
...
   */
 class CancellationSpec extends Fs2Suite {
-
-  case class Sub[A](b: AtomicBoolean) extends Subscriber[A] {
+  final class Sub[A](b: AtomicBoolean) extends Subscriber[A] {
     def onNext(t: A) = b.set(true)
     def onComplete() = b.set(true)
     def onError(e: Throwable) = b.set(true)
-    def onSubscribe(s: Subscription) = b.set(true)
+    def onSubscribe(s: Subscription) = ()
```
Comment (armanbilge) [inline reactive-streams/src/test/scala/fs2/interop/reactivestreams/CancellationSpec.scala:42]: Why this change?

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala [hunk 7]
```diff
@@ -25,35 +25,36 @@ package reactivestreams
...
 
 import org.reactivestreams._
 
+import scala.util.control.NoStackTrace
+
 /** Implementation of a `org.reactivestreams.Publisher`
   *
   * This is used to publish elements from a `fs2.Stream` to a downstream reactivestreams system.
   *
+  * @note Not longer unicast, this Publisher can be reused for multiple Subscribers.
```
[the remark this answers is not in the document]
Comment (armanbilge) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala:37]: Oh, and it's worth clarifying that the `Stream` will be compiled again for each subscription.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 8]
```diff
@@ -67,57 +68,87 @@ private[reactivestreams] final class StreamSubscription[F[_], A](
...
+      )
+    }.evalTap { subscription =>
+      F.delay(subscriber.onSubscribe(subscription))
+    }
+
+  def subscribe[F[_], A](stream: Stream[F, A], subscriber: Subscriber[A])(implicit
+      F: Async[F]
+  ): F[Unit] =
+    apply(stream, subscriber).use { subscription =>
+      F.onCancel(subscription.run, subscription.cancelMe)
```
[the remark this answers is not in the document]
Comment (BalmungSan) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:152]: I wondered the same when doing the implementation.
Not sure if there may be any use case for an uncancelable `run`

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 9]
```diff
@@ -23,35 +23,36 @@ package fs2
...
 )(implicit F: Async[F])
     extends Subscription {
   import StreamSubscription._
 
-  // We want to make sure `cancelled` is set _before_ signalling the subscriber
-  def onError(e: Throwable) = cancelled.set(true) >> F.delay(sub.onError(e))
-  def onComplete = cancelled.set(true) >> F.delay(sub.onComplete)
+  // Ensure we are on a terminal state before signaling the subscriber.
+  private def onError(e: Throwable): F[Unit] =
+    cancelMe >> F.delay(sub.onError(e))
```
[the remark this answers is not in the document]
Comment (BalmungSan) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:50]: _"terminal state"_ is the official term given the RX spec:

> For a **Publisher**: When `onComplete` or `onError` has been signalled.<br>
> For a **Subscriber**: When an `onComplete` or `onError` has been received.

I may merge both notes for extra clarity.

File: reactive-streams/src/test/scala/fs2/interop/reactivestreams/StreamUnicastPublisherSpec.scala [hunk 10]
```diff
@@ -23,35 +23,27 @@ package fs2
...
-      if (n == java.lang.Long.MAX_VALUE) Stream.range(1, 20).repeat
-      else Stream(1).repeat.scan(1)(_ + _).map(i => if (i > n) None else Some(i)).unNoneTerminate
+  override def createPublisher(n: Long): StreamUnicastPublisher[IO, Int] = {
+    val nums = Stream.constant(3)
 
-    StreamUnicastPublisher[IO, Int](s).allocated.unsafeRunSync()._1
+    StreamUnicastPublisher[IO, Int](
+      if (n == Long.MaxValue) nums else nums.take(n)
+    ).allocated.unsafeRunSync()._1
   }
```
[the remark this answers is not in the document]
Comment (BalmungSan) [reply reactive-streams/src/test/scala/fs2/interop/reactivestreams/StreamUnicastPublisherSpec.scala:41]: It was part of my process while debugging the test. TBH, it wasn't exactly clear how the old implementation was correct.
The `take` version is more straightforward and more evident IMHO.

File: reactive-streams/src/test/scala/fs2/interop/reactivestreams/CancellationSpec.scala (hunk 6 above)
Comment (BalmungSan) [reply reactive-streams/src/test/scala/fs2/interop/reactivestreams/CancellationSpec.scala:42]: `onSubscribe` must always be called, so if we set the _flag_ there the test would always fail.
Previously it worked because it wasn't even subscribing.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (armanbilge) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:152]: I am not sure I understand. `onCancel` doesn't make it `uncancelable`, it just sets up a finalizer.

[the remark this answers is not in the document]
Comment (BalmungSan) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:152]: Fair, in that sense I guess indeed the best is ensure `run` can be safely canceled.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 11]
```diff
@@ -68,14 +68,17 @@ private[reactivestreams] final class StreamSubscription[F[_], A] private (
...
+      stream
+        .through(subscriptionPipe)
+        .interruptWhen(canceled)
+        .evalMap(x => F.delay(sub.onNext(x)))
+        .handleErrorWith(e => Stream.eval(onError(e)))
+        .onFinalize(canceled.get.ifM(ifTrue = F.unit, ifFalse = onComplete))
+        .compile
+        .drain
+
+    F.onCancel(prog, cancelMe)
```
Comment (armanbilge) [inline reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:81]: I wonder if this should be a `.guarantee(...)`, since we probably want to set canceled no matter how we complete? I also wonder if it's redundant with the `onFinalize`.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 12]
```diff
@@ -67,57 +68,90 @@ private[reactivestreams] final class StreamSubscription[F[_], A](
...
       stream
         .through(subscriptionPipe)
-        .interruptWhen(cancelled)
+        .interruptWhen(canceled)
         .evalMap(x => F.delay(sub.onNext(x)))
         .handleErrorWith(e => Stream.eval(onError(e)))
-        .onFinalize(cancelled.get.ifM(ifTrue = F.unit, ifFalse = onComplete))
+        .onFinalize(canceled.get.ifM(ifTrue = F.unit, ifFalse = onComplete))
         .compile
         .drain
```
Comment (armanbilge) [inline reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:79]: Staring at this some more, I wonder if it should be:
```scala
      stream
        .through(subscriptionPipe)
        .foreach(x => F.delay(sub.onNext(x)))
        .compile
        .drain
        .guaranteeCase {
           case Success(...) => onComplete
           case Error(...) => onError
           case Canceled => onCancel
         }
        .race(canceled.get)
```
Then, I wonder if `canceled` can just be a `Deferred[Unit]` 🤔

**Review by armanbilge — COMMENTED** [review]
Overall looks great, one last round of bikeshed 😅

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala (hunk 11 above)
Comment (BalmungSan) [reply reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:85]: Yeah is redundant.

File: reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala [hunk 13]
```diff
@@ -68,14 +68,21 @@ private[reactivestreams] final class StreamSubscription[F[_], A] private (
...
+        .drain
+
+    val prog =
+      events.guaranteeCase {
+        case Outcome.Succeeded(_) => onComplete
+        case Outcome.Errored(ex)  => onError(ex)
+        case Outcome.Canceled()   => F.unit
+      }.race(canceled.get).void
+
+    F.onCancel(prog, cancelMe)
```
Comment (armanbilge) [inline reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscription.scala:85]: Shouldn't this be moved to the `Outcome.Canceled` case?
